### ISSUE TITLE
Extension Auto-Update Mechanism Enhancements

### ISSUE THREAD

**Issue by smcenlly** [microsoft/vscode#199482]
# Problem Statement:
Our team creates a [few popular VS Code extensions](https://marketplace.visualstudio.com/publishers/WallabyJs). Within our organisation, we have a continuous release model where it's not uncommon for us to release daily, and sometimes even multiple times each day. We've had a number of users who have expressed dissatisfaction with the current mechanism for reporting extension updates, particularly regarding their frequency and the annoyance of the notification system. This issue has been observed with several of our extensions, exhibited in poor reviews in our marketplace review page as well as issues in our issue tracking system.

We've also observed the same for other extensions expressing similar dissatisfaction, for example, [Github Copilot](https://marketplace.visualstudio.com/items?itemName=GitHub.copilot&ssr=false#review-details):

> **1 STAR**: Updates are too frequent. It raises questions about their Q/A process. Even ArchLinux has fewer updates than this.

The marketplace reviews are used to highlight the user sentiment / dissatisfaction.

We think that the key issues are:
1. Users believe (incorrectly) that a pending update means that they **must** restart their editor to continue using the extension with pending updates vs. continuing to use the old version until the next editor restart.
2. Users (especially those with diverse sensory / cognitive needs) are distracted by the pending updates badge on the Extensions side-bar badge and want to clear the pending notifications.

# Potential Solutions
We propose a possible set of improvements that could be implemented in part, or in whole. Where possible, we're happy to assist in contributing to improving this mechanism to expedite delivery (assuming it aligns with VS Code direction).

1. Allow extensions to opt-in for auto-update notifications (**not auto-updates**), either for a specific version of the extension, or for the extension entirely. This would allow extensions (both ours, and others) with continuous delivery models to not distract / annoy users. Perhaps a more general solution would be to notify for `major.minor` changes, but not `patch` releases (e.g. patch releases may auto-update without a badge indicator).
2. Allow the pending updates badge to be completely disabled for those with diverse sensory / cognitive needs.
3. Allow auto-update frequency and/or notification frequency to be controlled by users (e.g. every restart, once per day, once per week, etc.)

**Comment by smcenlly** [comment]
>If an extension has updates, user should be notified.

I think this is valid but only IF there is value in notifying them. If there are no release notes, or there is no actionable information for the user (other than to restart the editor), then the notification is just noise and interrupts their flow. I **think** for the majority of extensions (definitely for ours) that most updates are not critical and only affect a small subset of users (we notify these users to update in response to an issue / feature request).

> It should be up to users to opt out of notifications. I think the right approach here is to let users skip updates for this specific extension and also hide notification badge for this extension.

I still think that the providers of the extension should have the opportunity to have a say on this. For example, if there's a critical security update, then I think it's reasonable to notify users of that. I think it's also reasonable to notify users of updates that have actionable information for them (e.g. "we've added a new feature that you can use by doing X"). For our extensions, we do this via a custom webview with a LOT more information than we can fit in the release notes; we let users opt out of this. For example, most of our users don't care to know about a version upgrade for bleeding edge framework / node.js version support.

> May I know how often you publish your extension and what kind of changes go in?

The frequency of updates (and what's in them) very much depends on what we're working on. There is a baseload of constant updates to support latest versions of node, TypeScript, frameworks we integrate with, etc. On top of that, there are bigger features releases that are followed by incremental enhancements and bug fixes. Often times, the number of releases aligns with changes to the underlying tech we integrate with.

Here are the number of extension updates for **Quokka.js** 2023 YTD:

| Month | Number |
|-------|--------|
| Jan   | 5      |
| Feb   | 3      |
| Mar   | 2      |
| Apr   | 4      |
| May   | 9      |
| Jun   | 5      |
| Jul   | 2      |
| Aug   | 13     |
| Sep   | 4      |
| Oct   | 9      |
| Nov   | 23     |

To give you an idea of **some** of the November changes:

* Add support for `node@20.6.0+` new way of loading ESM hooks
* Fix `node@v21` JSDOM compatibility issue
* New Quokka Snaps feature
* Fix bug with latest version of TypeScript; ignore diagnostics report where Option 'module' must be set to 'NodeNext' when option 'moduleResolution' is set to 'NodeNext'
* Improve error reporting when ts-node transforms fail
* Update to improve re-using ts-node cache for transformations
* Fix bug in latest version of node with tsnode ESM integration
* Fix bug not setting jsxImportSource for TypeScript files
* Improve value output highlighting and formatting for string values
* Improve line mapping for transformed files

---

Having provided that information, as a team, we definitely (collectively) feel that there's a lot of room for improvement in VS Code's update mechanism, and that its too noisy right now. We are getting the same feedback from our users, and we see the same update patterns and feedback from users on other extensions. We also think this specific issue increases in severity based on the number of extensions a user has installed. If you have 10 extensions installed, then you're going to get 10x the number of updates than if you only have 1 extension. I wouldn't be surprised if many users are being notified of updates every day, and that's just unnecessary noise.

I also think that at some point, there's a straw that breaks the camel's back with updates, and users will remember the last update that happened to annoy them, and will complain about that one, even if there were 10 updates before that one that they didn't notice.

As mentioned above, if there's no actionable information available about an extension update for users, then the update notification is less useful. Sure, a user may need to update VS Code to apply the update, but if there's no value in the update to them, then the notification is just noise and their development flow has been interrupted.

I picked a few popular VS Code Extensions to show that the issue of frequent updates is not just related to our extensions. 

Here are some examples of other popular extensions that have frequent updates:

* [Python](https://marketplace.visualstudio.com/items?itemName=ms-python.python) - [details of changes](https://github.com/microsoft/vscode-python/releases) out of date (last updated on November 2, 2023)
  * 1/12/2023 7:49am
  * 30/11/2023 8:18pm
  * 30/11/2023 6:36am
  * 29/11/2023 8:15pm
  * 28/11/2023 8:19pm

* [Jupyter](https://marketplace.visualstudio.com/items?itemName=ms-toolsai.jupyter&ssr=false#version-history) - details of changes out of date (last mentioned via [VS Code release notes](https://code.visualstudio.com/updates/v1_84) in October 2023).
  * 29/11/2023 7:35pm 
  * 29/11/2023 11:00am
  * 29/11/2023 7:06am
  * 28/11/2023 7:24pm
  * 27/11/2023 7:24pm

* [GitLens](https://marketplace.visualstudio.com/items?itemName=eamodio.gitlens) - details of changes out of date (last mentioned November 21, 2023)
  * 30/11/2023 7:11pm
  * 29/11/2023 7:12pm
  * 28/11/2023 7:12pm
  * 27/11/2023 7:12pm
  * 24/11/2023 7:12pm

In addition to the GitHub Copilot example, some examples of negative feedback:

**Pylance Extension**
<img src="https://github.com/microsoft/vscode/assets/2074089/916da4e4-51e6-49cf-bfb2-c0321817e208" width="600" />

**JavaScript / TypeScript Nightly Build**
People even complain about the pre-release build...
<img src="https://github.com/microsoft/vscode/assets/2074089/e2ba9df5-33fc-4395-b54d-a507807d36d7" width="600" />

---

**In summary**: we feel (as evidenced above) that the default way that updates are displayed to users is too noisy. We feel that the VS Code UX needs to change to address this. We propose something like important vs. not-important updates. As extension authors, we would like to define which updates are important to display to users. For example, in terms of possible UX, not-important updates to an extension would not change the badge number, all updates would still be visible in the Extensions View.

**Comment by smcenlly** [comment]
@sandy081 - do you need any more information from us at this stage?

**Comment by sandy081** [comment]
@smcenlly Appreciated for detailed reasoning and feedback from users.

There are two kinds of notifications we have at present for extension updates

<img width="346" alt="image" src="https://github.com/microsoft/vscode/assets/10746682/f1d27b80-66d8-41e4-9acf-1a34205b32de">

1. Number badge on the Extensions Activity Bar - This is considered as prominent notification to the user for extension updates.
2. Action item on the extension in the view - This is considered as secondary information presentation. This is shown only when user opens this view. 

I tend to agree that the first mechanism should be able to consider the kind of extension transitions and should be configurable by the user. For eg., user can configure not to show update badge for a specific extension and then VS Code wont show the badge when there is an update for that extension. But when user opens the extensions view, VS Code should show the actions to update. VS Code should show the update badge for important updates like security fixes etc. This piece of information is missing that is to identify what kind of update it is.
